File: implementation/python/stepper.py
'''
A single-step reducer for tree calculus, on top of the representation in
`tree_calculus` (a tree is △ applied to its elements; applying one more argument
appends it to the tuple). Where `tree_calculus.reduce` runs a term to its normal
form in one go, `step` performs exactly one rewrite, so a reduction can be
observed term by term -- see `trace` / `trace_sampled`.
'''

from tree_calculus import format_term, parse_ternary


# --- rule 2, with optional peeks (--peek) ---
# The three shortcuts that avoid duplicating c into (x c) (b c): x = K yields
# c outright (b c is never built), and an eliminator (K f) as x or b collapses
# its application to f. Each is a contraction of the plain sequence, so normal
# forms agree; only the step granularity changes.

peek = False

def _apply(f, a):
    '''f applied to a, collapsing an eliminator: (K f) a -> f.'''
    return f[1] if len(f) == 2 and f[0] == () else f + (a,)

def _rule2(x, b, c, w):
    if not peek:
        return x + (c, b + (c,)) + w
    if x == ((),):                  # x = K: -> c
        return c + w
    return _apply(x, c) + (_apply(b, c),) + w

# --- single-step reduction (normal order, leftmost-outermost) ---
# `step` returns the next term, or None if `t` is already a normal form. It
# mirrors `apply` rule by rule, but stops after a single rewrite.
# `_fire` is the head rewrite itself, for a head whose rule is determined.

def _fire(t):
    a, b, c, *w = t
    w = tuple(w)
    if a == ():                     # rule 1: △ △ b c -> b
        return b + w
    if len(a) == 1:                 # rule 2: △ (△ x) b c -> (x c) (b c)
        (x,) = a
        return _rule2(x, b, c, w)
    x, y = a                        # rule 3: △ (△ x y) b c, by shape of c
    if c == ():                     # 3a: c = △     -> x
        return x + w
    if len(c) == 1:                 # 3b: c = △ u   -> (y u)
        (u,) = c
        return y + (u,) + w
    u, v = c                        # 3c: c = △ u v -> (b u v)
    return b + (u, v) + w

# --fuse: compress transient size spikes into one recorded composite -- after
# a fire that grew the head node, keep firing there, up to `fuse` fires total,
# while it is still above its pre-fire size. Every recorded term is a term of
# the plain sequence.

fuse = 0

def _fireable(t):
    return len(t) >= 3 and len(t[0]) < 3 and (len(t[0]) < 2 or len(t[2]) < 3)

def _fire_chain(t):
    out = _fire(t)
    start, n = nodes(t), 1
    while n < fuse and nodes(out) > start and _fireable(out):
        out = _fire(out); n += 1
    return out

def step(t):
    if len(t) >= 3:
        a, c = t[0], t[2]
        if len(a) >= 3:                 # function not a value yet: reduce it
            return (step(a),) + t[1:]
        if len(a) == 2 and len(c) >= 3: # rule-3 argument not a value yet
            return t[:2] + (step(c),) + t[3:]
        return _fire_chain(t)
    for i, e in enumerate(t):           # a value: reduce the leftmost child that can
        s = step(e)
        if s is not None:
            return t[:i] + (s,) + t[i+1:]
    return None

# --- applicative order (leftmost-innermost, an eager evaluator's discipline) ---
# Arguments are values before a rule fires, so values are duplicated rather
# than pending work: terms stay near the sizes an eager evaluator would see.
# Terms that are only weakly normalizing (the __wn family) diverge under this
# order; the default root-first `step` reaches their normal form.

def step_applicative(t):
    for i, e in enumerate(t):
        s = step_applicative(e)
        if s is not None:
            return t[:i] + (s,) + t[i+1:]
    return _fire(t) if len(t) >= 3 else None

# --- shrink-eager order (--shrink-eager) ---
# Only rule 2 with a non-leaf argument can grow a term; every other fire
# shrinks it without duplicating anything, so shrink redexes fire first --
# biggest drop first -- and only when nothing shrinks does the root-first
# `step` pick the growing fire. Firing a shrink early can only make every
# later term smaller (a residual argument), so intermediates stay small. Same
# normal forms; not proven normalizing, so pair with --limit if in doubt.
# Picking growing fires by any global rule (leftmost in preorder, min-growth)
# instead of root-first's descent provably diverges on parts of the __wn
# family.

def _drop(t):
    '''Node-count decrease of firing a shrink redex (context-free).'''
    a, b, c = t[0], t[1], t[2]
    if a == ():                     # rule 1 drops c
        return 2 + nodes(c)
    if len(a) == 1:                 # rule 2 on a leaf duplicates only the leaf
        return 1
    x, y = a
    if c == ():                     # 3a drops y and b
        return 3 + nodes(y) + nodes(b)
    if len(c) == 1:                 # 3b drops x and b
        return 3 + nodes(x) + nodes(b)
    return 3 + nodes(x) + nodes(y)  # 3c drops x and y

def step_shrink_eager(t):
    best = None                     # (drop, path) of the best shrink redex
    def walk(t, path):
        nonlocal best
        if (len(t) >= 3 and len(t[0]) < 3 and (len(t[0]) < 2 or len(t[2]) < 3)
                and (len(t[0]) != 1 or t[2] == ())):
            d = _drop(t)
            if best is None or d > best[0]:
                best = (d, path)
        for i, e in enumerate(t):
            walk(e, path + (i,))
    walk(t, ())
    if best is None:                # nothing shrinks: root-first's pick
        return step(t)
    def fire_at(t, path):
        if not path:
            return _fire(t)
        i = path[0]
        return t[:i] + (fire_at(t[i], path[1:]),) + t[i+1:]
    return fire_at(t, best[1])

def count_steps(t):
    n = 0
    while (s := step(t)) is not None:
        t = s; n += 1
    return n

def trace(t):
    '''Print every term in the reduction sequence; return the normal form.'''
    while True:
        print(format_term(t))
        s = step(t)
        if s is None:
            return t
        t = s

def nodes(t):
    '''Number of nodes: a term is △ applied to its elements, so one node plus
    the nodes of each element.'''
    return 1 + sum(map(nodes, t))

def trace_sampled(t, max_states=10, width=200):
    '''Like `trace`, but for long reductions: print only every k-th term (so at
    most `max_states` are shown), truncating each to `width` characters with its
    full length annotated. The final normal form is always printed.'''
    k = max(1, (count_steps(t) + 1 + max_states - 1) // max_states)
    def show(t):
        s = format_term(t)
        return s if len(s) <= width else f'{s[:width]}… ({len(s)} chars)'
    i = 0
    while True:
        s = step(t)
        if s is None:
            print(show(t))
            return t
        if i % k == 0:
            print(show(t))
        t = s; i += 1


if __name__ == '__main__':
    # Trace a reduction: stdin is whitespace-separated ternary terms, applied
    # left-associatively; each step prints "<step> <nodes>" (--terms appends
    # the full term in △ notation).
    import argparse, sys
    sys.setrecursionlimit(1_000_000)
    p = argparse.ArgumentParser(description='Trace a reduction step by step.')
    p.add_argument('--terms', action='store_true', help='also print each term')
    p.add_argument('--limit', type=int, help='stop after this many steps')
    p.add_argument('--eager', action='store_true', help='applicative instead of root-first order')
    p.add_argument('--peek', action='store_true', help='rule-2 shortcuts that avoid duplicating the argument')
    p.add_argument('--shrink-eager', action='store_true', help='shrinking fires first; smallest intermediate terms')
    p.add_argument('--fuse', type=int, nargs='?', const=8, default=0,
                   help='compress transient spikes into composite steps, up to this many fires (implies --peek)')
    a = p.parse_args()
    peek = a.peek or a.fuse > 0
    fuse = a.fuse
    # a fresh name: rebinding `step` would turn step_shrink_eager's root-first
    # fallback into a call to itself
    step_fn = step_applicative if a.eager else step_shrink_eager if a.shrink_eager else step
    t = ()
    for i, line in enumerate(sys.stdin.read().split()):
        u = parse_ternary(line)
        t = u if i == 0 else t + (u,)
    i = 0
    while True:
        print(f'{i} {nodes(t)} {format_term(t)}' if a.terms else f'{i} {nodes(t)}')
        s = step_fn(t)
        if s is None or (a.limit is not None and i >= a.limit):
            break
        t = s; i += 1

File: implementation/javascript/stepper.mjs
#!/usr/bin/env node
// One-rewrite-at-a-time reduction, fast enough to log every step of a long
// reduction: subterms are shared and carry cached node counts and a
// normal-form flag, so the whole-term size is cheap at every step.
//
// The default order is root-first (leftmost-outermost), which reaches the
// normal form of weakly normalizing terms -- at the price of duplicating
// pending work, which can grow terms an eager evaluator keeps small. --eager
// switches to eager order (arguments are values before a rule fires), where a
// zipper moves between redexes and a step is O(1) amortized.
//
// stdin: whitespace-separated ternary terms, applied left-associatively.
// Each step prints "<step> <nodes>"; --terms appends the term in △ notation,
// --limit N stops after N steps, --peek applies the rule-2 shortcuts that
// avoid duplicating the argument, --fuse [depth] (implies --peek) additionally
// compresses transient size spikes: after a fire that grew the head node, it
// keeps firing at that head -- up to depth (default 8) -- while the node is
// still above its pre-fire size, recording only the composite. Every recorded
// term is a term of the plain sequence, so the trace's peak is what a stepper
// with a depth-deep fused rule table would materialize.
//
// --shrink-eager is a different strategy over the plain rules (it composes
// with neither --peek nor --fuse): only rule 2 with a non-leaf argument can
// grow a term, and every other fire shrinks it without duplicating anything,
// so firing shrink redexes first -- biggest drop first, the growing rule 2
// leftmost-outermost only when nothing shrinks -- keeps every intermediate at
// most as large as under root-first order (a residual argument makes that
// pointwise). Same normal forms; each step is one canonical rewrite. Its
// leftmost pick is not proven normalizing, so pair it with --limit if in
// doubt (min-growth picks provably diverge on the __wn family).

import { readFileSync } from 'node:fs';

// A term is △ applied to its elements: { kids, n: nodes, nf: normal form }.
const LEAF = { kids: [], n: 1, nf: true };
const mk = kids => {
  let n = 1, nf = kids.length <= 2;
  for (const k of kids) { n += k.n; nf &&= k.nf; }
  return kids.length ? { kids, n, nf } : LEAF;
};

const patch = (node, idx, child) => {
  const kids = node.kids.slice();
  kids[idx] = child;
  return mk(kids);
};

// --peek: the three rule-2 shortcuts that avoid duplicating c into
// (x c) (b c) -- x = K yields c outright (b c is never built), and an
// eliminator (K f) as x or b collapses its application to f (apK).
// Contractions of the plain sequence: same normal forms, fewer steps.
const apK = (f, a) =>
  f.kids.length === 2 && f.kids[0].kids.length === 0 ? f.kids[1] : mk([...f.kids, a]);

// The rewrite at a spine head. Rules 1 and 2 never inspect c, so c may still
// hold pending work under root-first order; rule 3 requires its shape.
const fire = ([a, b, c, ...w]) => {
  if (a.kids.length === 0) return [...b.kids, ...w];                    // rule 1
  if (a.kids.length === 1) {                                            // rule 2
    const [x] = a.kids;
    if (!peek) return [...x.kids, c, mk([...b.kids, c]), ...w];
    if (x.kids.length === 1 && x.kids[0].kids.length === 0) return [...c.kids, ...w]; // x = K
    return [...apK(x, c).kids, apK(b, c), ...w];
  }
  const [x, y] = a.kids;                                                // rule 3, by shape of c
  if (c.kids.length === 0) return [...x.kids, ...w];
  if (c.kids.length === 1) return [...y.kids, c.kids[0], ...w];
  return [...b.kids, c.kids[0], c.kids[1], ...w];
};

const fireable = k =>
  k.length >= 3 && k[0].kids.length < 3 && (k[0].kids.length < 2 || k[2].kids.length < 3);

// One recorded rewrite: a single fire, or with --fuse the transient-skipping
// composite (see header).
const fireChain = k => {
  let out = fire(k);
  if (fuse) {
    const size = ks => ks.reduce((s, x) => s + x.n, 1);
    const start = size(k);
    for (let n = 1; n < fuse && size(out) > start && fireable(out); n++) out = fire(out);
  }
  return out;
};

const parseTernary = s => {
  const frames = [{ need: 1, kids: [] }];
  for (const ch of s) {
    frames.push({ need: +ch, kids: [] });
    while (frames.length > 1 && frames.at(-1).need === frames.at(-1).kids.length) {
      const f = frames.pop();
      frames.at(-1).kids.push(mk(f.kids));
    }
  }
  return frames[0].kids[0];
};

const fmt = root => {
  const out = [], stack = [root];
  while (stack.length) {
    const it = stack.pop();
    if (typeof it === 'string') { out.push(it); continue; }
    out.push('△');
    for (let i = it.kids.length - 1; i >= 0; i--) {
      const k = it.kids[i];
      if (k.kids.length === 0) stack.push(' △');
      else { stack.push(')', k, ' ('); }
    }
  }
  return out.join('');
};

const output = (() => {
  const buf = [];
  return (line, flush) => {
    if (line !== null) buf.push(line);
    if (buf.length === 65536 || (flush && buf.length)) {
      process.stdout.write(buf.join('\n') + '\n');
      buf.length = 0;
    }
  };
})();

// Root-first: fire at the outermost spine, descending only where a rule's
// shape dispatch demands a value (the function, and the argument of rule 3).
// Below the outermost redex, terms are values normalized left to right.
const stepNormal = root => {
  const path = [];
  let t = root;
  for (;;) {
    const k = t.kids;
    if (k.length >= 3) {
      if (k[0].kids.length >= 3) { path.push({ node: t, idx: 0 }); t = k[0]; continue; }
      if (k[0].kids.length === 2 && k[2].kids.length >= 3) { path.push({ node: t, idx: 2 }); t = k[2]; continue; }
      t = mk(fireChain(k));
      break;
    }
    const j = k.findIndex(x => !x.nf);
    if (j < 0) return null;                     // normal form
    path.push({ node: t, idx: j }); t = k[j];
  }
  for (let j = path.length - 1; j >= 0; j--) t = patch(path[j].node, path[j].idx, t);
  return t;
};

// One canonical rewrite, shrink-eager: the biggest-drop shrink redex anywhere,
// else the leftmost-outermost growing rule 2.
const shrinkDrop = k => {
  const [a, b, c] = k;
  if (a.kids.length === 0) return 2 + c.n;                              // rule 1
  if (a.kids.length === 1) return 1;                                    // rule 2, leaf c
  const [x, y] = a.kids;                                                // rule 3, by shape of c
  if (c.kids.length === 0) return 3 + y.n + b.n;
  if (c.kids.length === 1) return 3 + x.n + b.n;
  return 3 + x.n + y.n;
};
const stepShrinkEager = root => {
  let bestShrink = null, firstGrow = null;
  const walk = (t, path) => {
    if (t.nf) return;
    const k = t.kids;
    if (k.length >= 3 && k[0].kids.length < 3 && (k[0].kids.length < 2 || k[2].kids.length < 3)) {
      if (k[0].kids.length !== 1 || k[2].kids.length === 0) {
        const drop = shrinkDrop(k);
        if (!bestShrink || drop > bestShrink.drop) bestShrink = { path, drop };
      } else if (!firstGrow) firstGrow = { path };
    }
    for (let i = 0; i < k.length; i++) walk(k[i], path.concat(i));
  };
  walk(root, []);
  const target = bestShrink ?? firstGrow;
  if (!target) return null;
  const spine = [];
  let t = root;
  for (const i of target.path) { spine.push({ node: t, idx: i }); t = t.kids[i]; }
  t = mk(fire(t.kids));
  for (let j = spine.length - 1; j >= 0; j--) t = patch(spine[j].node, spine[j].idx, t);
  return t;
};

const traceNormal = (root, limit, terms, step = stepNormal) => {
  let i = 0;
  output(terms ? `${i} ${root.n} ${fmt(root)}` : `${i} ${root.n}`);
  while (i < limit && (root = step(root)) !== null) {
    i++;
    output(terms ? `${i} ${root.n} ${fmt(root)}` : `${i} ${root.n}`);
  }
  output(null, true);
};

const traceEager = (root, limit, terms) => {
  const frames = [];                    // { node, idx }: we are inside node.kids[idx]
  let focus = root, ctx = 0, i = 0;     // invariant: whole-term size = ctx + focus.n
  const rebuild = () => {
    let t = focus;
    for (let j = frames.length - 1; j >= 0; j--) t = patch(frames[j].node, frames[j].idx, t);
    return t;
  };
  const line = () =>
    output(terms ? `${i} ${ctx + focus.n} ${fmt(rebuild())}` : `${i} ${ctx + focus.n}`);
  line();
  while (i < limit) {
    while (focus.nf && frames.length) {         // done below here: rebuild upward
      const { node, idx } = frames.pop();
      ctx -= node.n - node.kids[idx].n;
      focus = patch(node, idx, focus);
    }
    if (focus.nf) break;
    for (;;) {                                  // descend to the leftmost-innermost redex
      const j = focus.kids.findIndex(k => !k.nf);
      if (j < 0) break;
      frames.push({ node: focus, idx: j });
      ctx += focus.n - focus.kids[j].n;
      focus = focus.kids[j];
    }
    focus = mk(fireChain(focus.kids));          // all kids normal, so arity >= 3
    i++; line();
  }
  output(null, true);
};

const args = process.argv.slice(2);
const li = args.indexOf('--limit');
const limit = li >= 0 ? Number(args[li + 1]) : Infinity;
const fi = args.indexOf('--fuse');
const fuse = fi >= 0 ? (Number(args[fi + 1]) || 8) : 0;
const peek = fuse > 0 || args.includes('--peek');
const toks = readFileSync(0, 'utf8').split(/\s+/).filter(Boolean);
let t = null;
for (const s of toks) {
  const u = parseTernary(s);
  t = t ? mk([...t.kids, u]) : u;
}
if (args.includes('--eager')) traceEager(t, limit, args.includes('--terms'));
else traceNormal(t, limit, args.includes('--terms'),
  args.includes('--shrink-eager') ? stepShrinkEager : stepNormal);
